src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+            {
+                actionsBuilder.Add(CreateNewCodeAction(FeaturesResources.Introduce_parameter_for_0,
+                    document, expression, methodSymbol, containingMethod, allOccurrences: false, trampoline: false, overload: false, cancellationToken));
+                actionsBuilder.Add(CreateNewCodeAction(FeaturesResources.Introduce_parameter_for_all_occurrences_of_0,
+                    document, expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: false, overload: false, cancellationToken));
+
+                if (!hasOptionalParameter && methodSymbol.MethodKind is not MethodKind.Constructor)
+                {
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_0, document, expression,
+                        methodSymbol, containingMethod, allOccurrences: false, trampoline: true, overload: false, cancellationToken));
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_all_occurrences_of_0, document,
+                        expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: true, overload: false, cancellationToken));
+
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Introduce_overload_with_new_parameter_for_0, document,
+                        expression, methodSymbol, containingMethod, allOccurrences: false, trampoline: false, overload: true, cancellationToken));
+                    actionsBuilder.Add(CreateNewCodeAction(
+                         FeaturesResources.Introduce_overload_with_new_parameter_for_all_occurrences_of_0, document,
+                         expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: false, overload: true, cancellationToken));
+                }
+            }
+
+            return actionsBuilder.ToImmutable();
+
+            // Local function to create a code action with more ease
+            MyCodeAction CreateNewCodeAction(string actionName, Document document,
+                TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+                bool allOccurrences, bool trampoline,
+                bool overload, CancellationToken cancellationToken)
+            {
+                return new MyCodeAction(actionName, c => IntroduceParameterAsync(
+                    document, expression, methodSymbol, containingMethod, allOccurrences, trampoline, overload, cancellationToken));
+            }
+        }
+
+        /// <summary>
+        /// Finds the matches of the expression within the same block.
+        /// </summary>
+        protected static async Task<IEnumerable<TExpressionSyntax>> FindMatchesAsync(Document originalDocument,
+            TExpressionSyntax expressionInOriginal, SyntaxNode withinNodeInCurrent,
```

Comment:
i don't understand what's going on here.  why not just find all the matches in the original as well.  why are you finding them 'in current'.

Reply:
withinNodeInCurrent is just the container method the expression was in

Reply:
probably poor naming, but withinNodeInCurrent just gets the scope, aka the parameterized method that it is located in. So it iterates through all expressionsyntax's in the scope

Reply:
i guess i don't understand the naming 'original' or 'current' here.  Nothing has changed right?  this is all just analyzing the starting document.  So this can just be `document/expression/withinNode`, right?

Reply:
yeah